### PR TITLE
Adding platform support for Juniper QFX5200

### PR DESCRIPTION
**Pull request by ciju-juniper**
This is a 1RU switch with 32 QSFP28 (40G/100G) ports on Broadcom Tomahawk I chipset. CPU used in QFX5200-32C-S is Intel Ivy Bridge. The switch has Redundant and hot-swappable Power Supply (1+1) and also has Redundant and hot swappable fans (5).

Signed-off-by: Ciju Rajan K <crajank@juniper.net>
Signed-off-by: Ashish Bhensdadia <bashish@juniper.net>

**- What I did**
- Added the device specific files, bcm configuration, portmapping file for Broadcom TH1
- Developed device drivers for platform specific fpga
- Developed the platform scripts

**- How I did it**
There is a issue https://github.com/Azure/sonic-buildimage/issues/4347 which is being debugged. This patch set has been verified after applying the workarounds/fixes provide on the latest of master branch.

**- How to verify it**
1. show platform summary
2. show interface status
3. show interface transceiver eeprom
4. sfputil show presence
5. sensors
6. show platform psustatus

The detailed UT was captured in the file qfx5200_unittest_logs.txt
[qfx5200_unittest_logs.txt](https://github.com/Azure/sonic-buildimage/files/4436480/qfx5200_unittest_logs.txt)

**- Description for the changelog**
Adding platform support for Juniper QFX5200

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by ciju-juniper** [comment]
@lguohan @jleveque Could you review and merge this platform?

**Comment by lguohan** [comment]
can you clean up the alerts raised by lgtm?

**Comment by ciju-juniper** [comment]
@lguohan Alerts are being cleaned up now. Shall upload the files soon.

**Comment by ciju-juniper** [comment]
@lguohan We have addressed all the LGTM alerts. Summary is not updated automatically as we edited in the pull request directly. Please consider merging the patchset.

**Comment by ciju-juniper** [comment]
@lguohan Now that  https://github.com/Azure/sonic-buildimage/pull/4379 has been committed, could you consider merging this platform?

**Comment by ciju-juniper** [comment]
@lguohan Rebased the patches to latest sonic-buildimage master.

**Comment by lguohan** [comment]
it looks like there are still some alerts from lgtm.

**Comment by ciju-juniper** [comment]
@lguohan A few alerts for nested exception handling. We had marked it as false positives ad these are valid scenarios.

2 Alerts are for "import *". This is needed right now.
2 Alerts are of variable defined. LGTB doesn't understand "os.command" so it's just flagging it as false positive.
Integer division alert is expected to have truncation. We are handling it already.

File: platform/broadcom/sonic-platform-modules-juniper/debian/files [hunk 1]
```diff
@@ -0,0 +1,3 @@
+sonic-juniper-platform-modules_1.1_amd64.buildinfo main extra
+sonic-platform-juniper-qfx5200_1.1_amd64.deb main extra
+sonic-platform-juniper-qfx5210_1.1_amd64.deb main extra
```
Comment (lguohan) [inline platform/broadcom/sonic-platform-modules-juniper/debian/files:3]: this looks like generated file, please remove.

File: platform/broadcom/sonic-platform-modules-juniper/qfx5200/modules/adt7470.c [hunk 2]
```diff
@@ -0,0 +1,1375 @@
...
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
```
Comment (lguohan) [inline platform/broadcom/sonic-platform-modules-juniper/qfx5200/modules/adt7470.c:19]: i can see linux has this driver. why not use the one from linux? 

https://github.com/torvalds/linux/blob/master/drivers/hwmon/adt7470.c

put into sonic-linux-kernel

Comment (ciju-juniper) [reply platform/broadcom/sonic-platform-modules-juniper/qfx5200/modules/adt7470.c:19]: @lguohan There are some platform specifc changes needed in QFX500 platform to make the adt7470 driver work. These changes will not be applicable for the other platforms. So we had to make a private copy of the driver adt7470.

Comment (ciju-juniper) [reply platform/broadcom/sonic-platform-modules-juniper/qfx5200/modules/adt7470.c:19]: @lguohan The default kernel driver adt7470 will not be able to adjust the fan speed in QFX5200, so we had to adjust a few parameters in the driver to make it functional in QFX5200.

File: platform/broadcom/sonic-platform-modules-juniper/debian/files (hunk 1 above)
Comment (ciju-juniper) [reply platform/broadcom/sonic-platform-modules-juniper/debian/files]: @lguohan Done.

**Comment by ciju-juniper** [comment]
@lguohan I've resolved your comments. Also provided the reason why we need adt7470 driver customization for QFX5200.

**Comment by ciju-juniper** [comment]
@lguohan We re-organized the code and fixed all the lgtm python alerts. Requesting you to merge this platform.

**Comment by ciju-juniper** [comment]
@lguohan Could you merge this pull request?

**Comment by ciju-juniper** [comment]
retest vsimage

**Comment by ciju-juniper** [comment]
@lguohan  We have tested on the latest of master and everything is working as expected. Please do merge the patch set.

**Comment by jleveque** [comment]
Retest vsimage please

**Comment by ciju-juniper** [comment]
@jleveque I see that you were able to trigger the vsimage test. I tried it earlier by adding 'retest vsimage' comment, but it didn't trigger. Does it mean only maintainers can trigger it?

I see that last test in vsimage test suite is always failing with most of the pull requests. Are there any issues with the test suite? At least this pull request doesn't touch any of the VS platform files.

**Comment by jleveque** [comment]
@ciju-juniper: You can also trigger it. The plugin requires you add the word "please" at the end (it forces you to be polite :) ).

I'm not aware of issues with the vsimage test suite, but there may be. I'll see if anyone is aware of this.

**Comment by ciju-juniper** [comment]
@jleveque Thanks.. I will be more polite from now on ;-)

**Comment by ciju-juniper** [comment]
@jleveque Would you be able to merge this pull request? There aren't any changes which affects vsimage.

**Comment by jleveque** [comment]
I will wait for @lguohan to approve and make the decision.

File: platform/broadcom/sonic-platform-modules-juniper/qfx5200/modules/adt7470.c (hunk 2 above)
Comment (lguohan) [reply platform/broadcom/sonic-platform-modules-juniper/qfx5200/modules/adt7470.c:19]: what is the nature of these private changes? why you won't be able to make it as a generic kernel patch?

Comment (ciju-juniper) [reply platform/broadcom/sonic-platform-modules-juniper/qfx5200/modules/adt7470.c:19]: @lguohan We had a patch for adjusting the PWM frequency using the 0x40 register. Initially @ashishb-juniper  added it for QFX5200 and now I see that it's already in the mainline: https://patchwork.kernel.org/patch/9266607/

I've removed the private driver. Also tested the switch by loading the default adt7470 driver. Updating the pull request.

**Comment by ciju-juniper** [comment]
@lguohan I've refreshed the pull request after removing the adt7470 private driver.

**Comment by jleveque** [comment]
@ciju-juniper: Please try to use force-pushes only when necessary. Having a history of commits in GitHub make reviewing iterations easier.

**Comment by ciju-juniper** [comment]
@jleveque I will keep that in mind. I was trying to get a single clean commit in the pull request as it's a new platform addition.

**Comment by jleveque** [comment]
@ciju-juniper: You don't need to worry about getting a single, clean commit. When we merge pull requests, we use a "squash merge" which will squash all commits in the PR into one single, clean commit.

**Comment by ciju-juniper** [comment]
@jleveque Thanks for the info.

**Comment by ciju-juniper** [comment]
@lguohan @jleveque  Looks like all the builds/tests passed this time. Would you be able to merge now? Also, I've addressed your comment regarding adt7470 driver.

**Comment by ciju-juniper** [comment]
@lguohan @jleveque I see that all the checks have passed now including the review. Would this pull request be merged automatically?

**Pull request merged**